src/server/main_service.cc

```diff
@@ -1433,15 +1433,22 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
     sinfo->keys.insert(KeyLockArgs::GetLockKey(ArgS(eval_args.keys, i)));
   }
   sinfo->async_cmds_heap_limit = absl::GetFlag(FLAGS_multi_eval_squash_buffer);
-  DCHECK(cntx->transaction);
+  Transaction* tx = cntx->transaction;
+  CHECK(tx != nullptr);
 
-  bool scheduled = StartMultiEval(cntx->db_index(), eval_args.keys, *params, cntx->transaction);
+  bool scheduled = StartMultiEval(cntx->db_index(), eval_args.keys, *params, tx);
 
   interpreter->SetGlobalArray("KEYS", eval_args.keys);
   interpreter->SetGlobalArray("ARGV", eval_args.args);
   interpreter->SetRedisFunc([cntx, this](auto args) { CallFromScript(cntx, args); });
 
-  Interpreter::RunResult result = interpreter->RunFunction(eval_args.sha, &error);
+  Interpreter::RunResult result;
+  if (tx->GetMultiMode() == Transaction::LOCK_AHEAD && tx->GetUniqueShardCnt() == 1) {
+    ShardId sid = tx->GetUniqueShard();
+    pp_.at(sid)->Await([&]() { result = interpreter->RunFunction(eval_args.sha, &error); });
```

Comment:
I just remembered: if we move ove the execution to another thread, don't we access the connection context from another thread as well? We need to be safe in that case because we might have a dispatch fiber still working with it

Reply:
Wow, that's an excellent comment.
In what circumstances could there be other fibers active? In the case of replication, maybe?
It sounds like a hard problem to solve (pausing or waiting for other fibers because moving execution). Do you have any advice?

Reply:
> In what circumstances could there be other fibers active? In the case of replication, maybe?

In case of pipelined execution, one fiber reads, the other executes, they communicate over one queue: the `dispatch_q` in code

---

It is unsafe now in some cases. For example, if the monitor is enabled, monitor messages will be added to the dispatch queue from this other thread of execution. If this EVAL command was part of a pipeline, then the dispatch queue is accessed by the reader fiber at the same time -> race

That was about EVAL only ⬆️, not the commands it executes. If you execute any commands that access the context mutably, it might break as well.

---

How does squashing work? Squashing creates context copies and captures replies to re-order them safely. Other stuff like monitoring or even mutating the context doesn't work there from non-native threads.

--- 

> Do you have any advice?

Not yet. I'm working on this (that's why I remembered 😁 ) - but for all commands to parallelize pipelines. Once I solved it for my case, it'll be solved for this as well